torch/csrc/distributed/c10d/symm_mem/nccl_devcomm_manager.hpp

```diff
@@ -0,0 +1,90 @@
+#pragma once
+
+#ifdef NCCL_HAS_SYMMEM_DEVICE_SUPPORT
+
+#include <c10/cuda/CUDAException.h>
+#include <c10/cuda/CUDAGuard.h>
+#include <c10/util/Exception.h>
+#include <nccl_device.h>
+#include <string>
+#include <unordered_map>
+
+namespace c10d::symmetric_memory {
+
+// Manage all the NCCL device communicator business. Singleton.
+class NCCLDevCommManager {
+ public:
+  // Constructor
+  explicit NCCLDevCommManager(const c10::Device device) : device_(device) {}
+
+  // Get single, global manager.
+  static NCCLDevCommManager& get(const c10::Device device) {
+    static NCCLDevCommManager manager(device);
+    TORCH_CHECK(
+        manager.device_ == device,
+        "Detected use of NCCLDevCommManager on multiple devices. This is not supported.");
+    return manager;
+  }
+
+  // Get an NCCL device communicator for a group.
+  ncclDevComm get_devcomm(const std::string& group_name) {
+    auto it = group_to_comms_.find(group_name);
+    if (it == group_to_comms_.end()) {
+      TORCH_CHECK(
+          false,
+          "NCCL device communicator for group ",
+          group_name,
+          " not found");
+    }
+    return it->second.second;
+  }
+
+  // Create device communicator if it doesn't exist. Skip if it already exists.
+  void try_emplace_devcomm(const std::string& group_name, ncclComm_t comm) {
```

Comment:
You are doing try_emplace, any reason not to use perfect forwarding?

Reply:
The comm passed here is a host comm. Need to call `ncclDevCommCreate` to create a device one. But that function is an in-place func.